File: ts/test/resolved.test.ts
/* Copyright (c) 2024-2026 Voxgig Ltd, MIT License */

import { test, describe } from 'node:test'
import assert from 'node:assert'

import { parse } from '../dist/apidef'
import { contractTransform } from '../dist/transform/contract'
import { operationFacts, operationIndex, makeResolved, publishResolved, resolvedSpec }
  from '../dist/resolved'


const SPEC = {
  openapi: '3.0.0',
  info: { title: 't', version: '1' },
  security: [{ apiKeyAuth: [] }],
  components: { securitySchemes: { apiKeyAuth: { type: 'http', scheme: 'bearer' } } },
  paths: {
    '/things/{id}': {
      parameters: [{ in: 'path', name: 'id', required: true, schema: { type: 'string' } }],
      get: {
        operationId: 'getThing',
        parameters: [{ in: 'query', name: 'expand', schema: { type: 'string' } }],
        responses: { 200: { description: 'ok' } },
      },
      delete: { security: [], responses: { 204: { description: 'gone' } } },
    },
    '/open': { get: { responses: { 200: { description: 'ok' } } } },
  },
}


describe('resolved', () => {

  test('merges path-level and operation-level parameters', () => {
    const facts: any = operationFacts(SPEC, { method: 'GET', orig: '/things/{id}' })
    assert.deepEqual(facts.parameters.map((p: any) => p.name), ['id', 'expand'])
  })

  test('defaults security from the document and says where it came from', () => {
    const get: any = operationFacts(SPEC, { method: 'GET', orig: '/things/{id}' })
    assert.deepEqual(get.security, [{ apiKeyAuth: [] }])
    assert.equal(get.securitySource, 'definition')

    // An operation that overrides with [] needs NO auth — a distinction the
    // model's single resolved `kit.info.security` cannot express.
    const del: any = operationFacts(SPEC, { method: 'DELETE', orig: '/things/{id}' })
    assert.deepEqual(del.security, [])
    assert.equal(del.securitySource, 'operation')
  })

  test('carries securitySchemes under either specification spelling', () => {
    const oas3: any = operationFacts(SPEC, { method: 'GET', orig: '/open' })
    assert.deepEqual(Object.keys(oas3.securitySchemes), ['apiKeyAuth'])

    const swagger2: any = operationFacts({
      swagger: '2.0',
      securityDefinitions: { basic: { type: 'basic' } },
      paths: { '/x': { get: { responses: {} } } },
    }, { method: 'GET', orig: '/x' })
    assert.deepEqual(Object.keys(swagger2.securitySchemes), ['basic'])
  })

  test('is undefined for an operation the definition does not describe', () => {
    assert.equal(operationFacts(SPEC, { method: 'PUT', orig: '/things/{id}' }), undefined)
    assert.equal(operationFacts(SPEC, { method: 'GET', orig: '/nope' }), undefined)
  })

  test('indexes every described operation by method and path', () => {
    assert.deepEqual(Object.keys(operationIndex(SPEC)).sort(),
      ['DELETE /things/{id}', 'GET /open', 'GET /things/{id}'])
  })

  test('publishes onto the shared build context, and reads back', () => {
    const buildctx: any = { step: 'pre', state: {} }
    publishResolved(buildctx, 'openapi3', SPEC)

    // @voxgig/model mutates step on ONE context object, so what the pre step
    // publishes is what the post step reads.
    buildctx.step = 'post'
    const back = resolvedSpec(buildctx)
    assert.equal(back?.kind, 'openapi3')
    assert.equal(back?.operation('GET', '/things/{id}')?.operationId, 'getThing')
  })

  test('publishing without a build context still returns the capability', () => {
    // apidef also runs outside a model build.
    const r = publishResolved(undefined, 'openapi3', SPEC)
    assert.equal(r.operation('GET', '/open')?.protocol, 'http')
    assert.equal(resolvedSpec(undefined), undefined)
  })

  test('the definition it carries is the PARSED one, not the file', async () => {
    // A colon-style path is normalised during parse; a consumer reading the
    // raw file would miss every lookup on such a specification.
    const raw = JSON.stringify({
      openapi: '3.0.0', info: { title: 't', version: '1' },
      paths: {
        '/projects/:slug': {
          get: { parameters: [{ in: 'path', name: 'slug' }], responses: {} },
        },
      },
    })
    const def: any = await parse('OpenAPI', raw, { file: 'test.json' })
    const r = makeResolved('openapi3', def)
    assert.equal(r.operation('GET', '/projects/{slug}')?.protocol, 'http')
    assert.equal(r.operation('GET', '/projects/:slug'), undefined)
  })

})


// The guide's hint, not a specification fact, so it must survive on the point.
describe('live-hint-on-point', () => {

  test('a guide live hint lands on the point', async () => {
    const point: any = { method: 'GET', orig: '/things', kind: 'http' }
    const ctx: any = {
      def: { paths: { '/things': { get: { responses: {} } } } },
      apimodel: { main: { kit: { entity: { thing: { name: 'thing',
        op: { list: { name: 'list', points: [point] } } } } } } },
      guide: { entity: { thing: { path: { '/things': { op: { list: { live: true } } } } } } },
    }

    await contractTransform(ctx)

    assert.equal(point.live, true)

    assert.equal(point.contract.json, undefined)
    assert.equal(point.contract.id, 'GET /things')
  })

  test('no hint leaves the point alone', async () => {
    const point: any = { method: 'GET', orig: '/things', kind: 'http' }
    const ctx: any = {
      def: { paths: { '/things': { get: { responses: {} } } } },
      apimodel: { main: { kit: { entity: { thing: { name: 'thing',
        op: { list: { name: 'list', points: [point] } } } } } } },
      guide: {},
    }

    await contractTransform(ctx)

    assert.equal(point.live, undefined)
  })

})

File: go/resolved_test.go
package apidef

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

func TestResolvedPublication(t *testing.T) {
	schema := map[string]any{"type": "object"}
	schema["properties"] = map[string]any{"child": schema}
	def := map[string]any{"paths": map[string]any{"/items": map[string]any{"post": map[string]any{"requestBody": schema}}}}
	guide := map[string]any{}
	ctx := map[string]any{"state": map[string]any{"apidef": map[string]any{"retained": true}}}
	resolved := PublishResolved(ctx, "openapi3", def, func() map[string]any { return guide })
	guide = map[string]any{"entity": map[string]any{"item": map[string]any{"path": map[string]any{"/items": map[string]any{"op": map[string]any{
		"create": map[string]any{"method": "POST", "contract": map[string]any{"security": []any{}}, "live": false},
	}}}}}}
	facts, err := resolved.Operation("POST", "/items")
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(facts["security"], []any{}) || facts["live"] != false {
		t.Fatalf("guide lost: %v", facts["security"])
	}
	if reflect.ValueOf(facts["requestBody"]).Pointer() != reflect.ValueOf(schema).Pointer() {
		t.Fatal("schema copied")
	}
	for _, carrier := range []any{ctx, &ApiDefResult{Ctx: &ApiDefContext{Resolved: resolved}}, map[string]any{"ctx": ctx}, map[string]any{"apidef": map[string]any{"resolved": resolved}}} {
		if ResolvedSpecFrom(carrier) != resolved {
			t.Fatal("capability not found")
		}
	}
	if ctx["state"].(map[string]any)["apidef"].(map[string]any)["retained"] != true {
		t.Fatal("state overwritten")
	}
	if PublishResolved(nil, "openapi3", def) == nil || ResolvedSpecFrom(nil) != nil {
		t.Fatal("missing context")
	}
	index := OperationIndex(def)
	if len(index) != 1 || index["POST /items"] == nil {
		t.Fatal("operation index")
	}
}

func TestResolvedGenerate(t *testing.T) {
	dir := t.TempDir()
	if err := os.MkdirAll(filepath.Join(dir, "def"), 0755); err != nil {
		t.Fatal(err)
	}
	raw := `{"openapi":"3.0.0","info":{"title":"t","version":"1"},"servers":[{"url":"https://example.com"}],"paths":{"/items/{id}":{"get":{"parameters":[{"name":"id","in":"path","required":true,"schema":{"type":"string"}}],"responses":{"200":{"description":"ok"}}}}}}`
	if err := os.WriteFile(filepath.Join(dir, "def", "test.json"), []byte(raw), 0644); err != nil {
		t.Fatal(err)
	}
	for _, stop := range []string{"guide", "transformers", "builders", "generate"} {
		t.Run(stop, func(t *testing.T) {
			buildctx := map[string]any{}
			result, err := NewApiDef(ApiDefOptions{Folder: stageGuideEntry(t, t.TempDir(), "")}).Generate(map[string]any{
				"model":    map[string]any{"name": "test", "def": "test.json"},
				"build":    map[string]any{"spec": map[string]any{"base": filepath.Join(dir, "model")}},
				"buildctx": buildctx, "ctrl": map[string]any{"step": map[string]any{stop: false}},
			})
			if err != nil {
				t.Fatal(err)
			}
			resolved := ResolvedSpecFrom(result)
			if resolved == nil || ResolvedSpecFrom(buildctx) != resolved {
				t.Fatal("pipeline did not publish capability")
			}
			facts, err := resolved.Operation("GET", "/items/{id}")
			if err != nil || facts == nil {
				t.Fatalf("parsed operation missing: %v", err)
			}
			if reflect.ValueOf(resolved.Def).Pointer() != reflect.ValueOf(result.Ctx.Def).Pointer() {
				t.Fatal("capability copied the parsed definition")
			}
		})
	}
}

func TestModelPointLiveRoundTrip(t *testing.T) {
	for _, src := range []string{`{"li":true}`, `{"li":false}`, `{"li":{"input":{"n":2}}}`} {
		var point ModelPoint
		if err := json.Unmarshal([]byte(src), &point); err != nil {
			t.Fatal(err)
		}
		data, err := json.Marshal(point)
		if err != nil {
			t.Fatal(err)
		}
		var got, want map[string]any
		json.Unmarshal(data, &got)
		json.Unmarshal([]byte(src), &want)
		if !reflect.DeepEqual(got["li"], want["li"]) {
			t.Fatalf("live lost: %s", data)
		}
	}
}
